From 739a1bbe975bb31543167d4eab65e6f9b9cdb6e7 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Wed, 29 May 2024 19:16:30 +0300
Subject: [PATCH] [FSW-420] fs1: image security validation

with this patch, added spi manifest validation checks
of public key, sha, crc32 and firmware modules
sha verification.

Change-Id: I3d0876bb8126a22d7815afa9a744688361fec036
---

diff --git a/app/boot_rom.c b/app/boot_rom.c
index b024b62..a63422e 100644
--- a/app/boot_rom.c
+++ b/app/boot_rom.c
@@ -6,6 +6,7 @@
 
 #include <boot_rom.h>
 #include <security.h>
+#include <recovery.h>
 
 struct boot_rom_global_s __attribute__((section (".rom_persistent_section")))boot_rom_global = {
 	.config_fuse = 0,
@@ -33,6 +34,9 @@
 
 	boot_rom_check_n_enter_recovery();
 
+	if (boot_rom_verify_image(HL_FW_BOOT_PTR_SPI))
+		execute_recovery();
+
 	boot_rom_bringup_ppboot();
 
 	return 0;
diff --git a/arch/fs1/arc/Makefile b/arch/fs1/arc/Makefile
index 22879ed..35c7378 100644
--- a/arch/fs1/arc/Makefile
+++ b/arch/fs1/arc/Makefile
@@ -12,3 +12,5 @@
 BOOT_ROM_COMMON_SRCS_C += $(shell ls $(HL_DEV_ARCH)/dev_uart.c)
 BOOT_ROM_COMMON_SRCS_C += $(shell ls $(HL_DEV_ARCH)/dev_recovery.c)
 BOOT_ROM_COMMON_SRCS_C += $(shell ls $(HL_DEV_ARCH)/dev_dma.c)
+BOOT_ROM_COMMON_SRCS_C += $(shell ls $(HL_DEV_ARCH)/dev_security.c)
+BOOT_ROM_COMMON_SRCS_C += $(shell ls $(HL_DEV_ARCH)/security.c)
diff --git a/arch/fs1/arc/dev_security.c b/arch/fs1/arc/dev_security.c
new file mode 100644
index 0000000..5416c1f
--- /dev/null
+++ b/arch/fs1/arc/dev_security.c
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
+
+#include <mem_io.h>
+#include <string.h>
+#include <boot_rom.h>
+#include "dev_security.h"
+#include <fs1_blocks_linux_driver.h>
+#include <fs1_reg_map.h>
+#include <psoc_security_regs.h>
+
+/*
+ * This key hash is calculated on the Public Key Properties -
+ *   -----------------------------------------------
+ * | R ^ 2 | modulus | exponent | inverse | num bits |
+ *   -----------------------------------------------
+ * Since the key props are directly given to the ROM, there is
+ * no need to explicitly include the Public key, which is
+ * usually the modulus.
+ */
+const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN] = {
+	/* Habana Public Key */
+	{
+		0x95, 0xa3, 0x2a, 0x03, 0xd4, 0xbd, 0xb1, 0xe0,
+		0x2b, 0x54, 0xc0, 0x9c, 0x61, 0x10, 0x80, 0x2c,
+		0x08, 0xc4, 0x5d, 0xdf, 0x4a, 0xf8, 0x03, 0x00,
+		0x2a, 0xf5, 0xac, 0x20, 0x6f, 0x2c, 0x8f, 0x27,
+		0x69, 0x2e, 0x63, 0x7d, 0x02, 0x33, 0x54, 0x29,
+		0x3e, 0x57, 0x46, 0xd9, 0xf4, 0xa9, 0x7e, 0x88
+	},
+	/* Production Key 1 */
+	{
+		0x60, 0x44, 0x16, 0xd8, 0x1b, 0xea, 0x2e, 0x66,
+		0x46, 0xf3, 0x01, 0x78, 0x2f, 0x3c, 0xf6, 0xfc,
+		0x88, 0x3d, 0xbb, 0x5d, 0x94, 0xe7, 0xd9, 0x94,
+		0x79, 0xea, 0x6e, 0xdc, 0x1d, 0xd8, 0x89, 0xe1,
+		0xa4, 0xeb, 0x28, 0x4a, 0x1b, 0x60, 0x87, 0xed,
+		0x58, 0x4a, 0xe1, 0xa0, 0xe9, 0x13, 0x6f, 0xb6
+	},
+	/* Production Key 2 */
+	{
+		0x0c, 0xbe, 0x41, 0xaa, 0x56, 0xd2, 0xe3, 0x29,
+		0x0f, 0x45, 0x2f, 0x48, 0x64, 0x1f, 0x61, 0x00,
+		0xa3, 0x94, 0x6c, 0x2b, 0xc0, 0xa5, 0xd0, 0x9a,
+		0x0d, 0x3d, 0xb3, 0x74, 0x85, 0x1c, 0xbc, 0x9b,
+		0xee, 0xc1, 0x3c, 0xbc, 0x7e, 0xae, 0xf0, 0x7c,
+		0x4a, 0x7e, 0x82, 0x52, 0x69, 0xbc, 0xdc, 0x5f
+	},
+	/* Production Key 3 */
+	{
+		0xd3, 0x21, 0x9d, 0xe5, 0xcf, 0x7e, 0x4c, 0x20,
+		0x98, 0x4e, 0x84, 0x4d, 0x00, 0x5f, 0xf8, 0x71,
+		0x98, 0xf6, 0x11, 0x8f, 0xd4, 0xfd, 0xf6, 0x29,
+		0x6a, 0xd3, 0xfd, 0x94, 0xcf, 0x21, 0xf5, 0x60,
+		0xe8, 0x78, 0x91, 0x4e, 0xc0, 0x51, 0xc6, 0x82,
+		0xe1, 0xae, 0xfe, 0xcf, 0xbf, 0x96, 0xcf, 0xaf
+	},
+	/* Production Key 4 */
+	{
+		0x98, 0xd0, 0x0c, 0xa3, 0xa4, 0xc3, 0xe8, 0x55,
+		0x8d, 0xcf, 0xd2, 0x48, 0x0d, 0x8a, 0x52, 0xd6,
+		0xf5, 0x4c, 0x45, 0x58, 0x6d, 0x20, 0xb9, 0xe2,
+		0x4f, 0xe7, 0x6d, 0x0b, 0xdc, 0x76, 0xf6, 0xad,
+		0x2f, 0x79, 0xca, 0x03, 0x92, 0xc9, 0x2a, 0x5f,
+		0x36, 0x19, 0xee, 0x61, 0x3a, 0x10, 0x69, 0xa9
+	}
+};
+
+void security_report_img_status(uint8_t index, uint32_t img_sts_bit)
+{
+	uint32_t reg, offset;
+
+	if (index == BOOT_ROM_SECONDARY)
+		offset = IO0_H0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_SEC_IMG_STATUS);
+	else
+		offset = IO0_H0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_PRI_IMG_STATUS);
+
+	reg = regmap_readl(offset);
+	reg |= img_sts_bit;
+	regmap_writel(reg, offset);
+}
diff --git a/arch/fs1/arc/dev_security.h b/arch/fs1/arc/dev_security.h
new file mode 100644
index 0000000..69c28ef
--- /dev/null
+++ b/arch/fs1/arc/dev_security.h
@@ -0,0 +1,32 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+#ifndef _DEV_SECURITY_H_
+#define _DEV_SECURITY_H_
+
+#define NUM_KEYS_IN_ROM					5
+#define SECURITY_MAX_FW_SEC_VERSION		128
+
+/* size of key properties in bytes */
+#define KEY_PROP_RR_SZ					384
+#define KEY_PROP_MOD_SZ					384
+#define KEY_PROP_EXP_SZ					8
+#define KEY_PROP_INV_SZ					4
+#define KEY_PROP_NB_SZ					4
+#define KEY_PROP_HDR_SZ					8
+
+/* 384 + 384 + 8 + 4 + 4 = 784 bytes */
+#define PUB_KEY_SZ		(KEY_PROP_RR_SZ + KEY_PROP_MOD_SZ + \
+				KEY_PROP_EXP_SZ + KEY_PROP_INV_SZ + \
+				KEY_PROP_NB_SZ)
+
+/* 784 + 8 = 792 */
+#define TOTAL_PUB_KEY_PROPS_SZ  (PUB_KEY_SZ + KEY_PROP_HDR_SZ)
+
+void security_report_img_status(uint8_t index, uint32_t img_sts_bit);
+
+extern const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN];
+
+#endif /* _DEV_SECURITY_H_ */
diff --git a/arch/fs1/arc/soc.h b/arch/fs1/arc/soc.h
index 136cc87..22159e7 100644
--- a/arch/fs1/arc/soc.h
+++ b/arch/fs1/arc/soc.h
@@ -32,6 +32,8 @@
 	GPIO_FUNC_3,
 };
 
+#define SIG_SIZE			384
+
 #define REGION_BASE_ADDR(REG_IDX)	(REG_IDX << 28)
 #define SZ_32K				0x00008000
 #define SZ_64K				0x00010000
diff --git a/common/boot_desc.c b/common/boot_desc.c
index bc4ed9e..0490d75 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -229,3 +229,18 @@
 
 	return BOOT_ROM_IMAGE_NOT_FOUND;
 }
+
+uint32_t boot_desc_param_get(enum hl_fw_boot_ptr_imgs img, struct hl_fw_boot_ptrs *boot_ptr)
+{
+	boot_desc_update_ptrs(img);
+
+	if (!p_boot_ptr)
+		return 0;
+
+	memcpy(boot_ptr, p_boot_ptr, sizeof(struct hl_fw_boot_ptrs));
+
+	if (img == HL_FW_BOOT_PTR_RECOVERY)
+		return BOOT_PTR_REC_DMA_DEST_ARC;
+	else
+		return BOOT_PTR_SPI_DMA_DEST_ARC;
+}
diff --git a/common/boot_desc.h b/common/boot_desc.h
index b2f20d9..670880f 100644
--- a/common/boot_desc.h
+++ b/common/boot_desc.h
@@ -29,6 +29,6 @@
 int boot_desc_init(enum hl_fw_boot_ptr_imgs img);
 uint32_t boot_desc_get_img_desc(enum hl_fw_boot_ptr_imgs img, int type,
 				struct boot_desc *desc, enum fw_post_reset_cmd upd_status);
+uint32_t boot_desc_param_get(enum hl_fw_boot_ptr_imgs img, struct hl_fw_boot_ptrs *boot_ptr);
 
 #endif /* _BOOT_DESC_H_ */
-
diff --git a/common/common.c b/common/common.c
index 862456d..3fb5521 100644
--- a/common/common.c
+++ b/common/common.c
@@ -16,6 +16,9 @@
 #include "dev_uart.h"
 #include <boot_desc.h>
 #include <dma.h>
+#include <dev_security.h>
+#include <secure_boot.h>
+#include <psoc_security_masks.h>
 
 void boot_rom_uart_init(void)
 {
@@ -156,6 +159,10 @@
 	enum boot_rom_progress img_sts_base = PROGRESS_BRINGUP_PCI_PRI;
 	boot_rom_global.curr_img = cur_img;
 
+	uint32_t act_img_size, loadaddr_arc;
+	int ret;
+	unsigned char *image;
+
 	if (cur_img == BOOT_ROM_SECONDARY)
 		img_sts_base = PROGRESS_BRINGUP_PCI_SEC;
 		/* TODO: Check what is TPM_SPI and progress status */
@@ -207,8 +214,26 @@
 	}
 
 	boot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_DMA);
+	
+	if (!security_efuse_image_validation_en())
+		return BOOT_ROM_SUCCESS;
 
-	/* TODO: security_impl */
+	act_img_size = desc->img_size;
+	act_img_size = act_img_size - CRC32_SIZE;
+
+	/* Derive key and signature parameters */
+	loadaddr_arc = (uint32_t)boot_rom_get_arc_addr(desc->loadaddr);
+	image = (unsigned char *)(loadaddr_arc);
+	//image_len = act_img_size;
+
+	ret = secureboot_verify_image_hash384(desc->chksum, image, act_img_size);
+	if (ret) {
+		if (img != HL_FW_IMG_TPM_SPI)
+			security_report_img_status(cur_img, PSOC_SECURITY_PRI_IMG_STATUS_AUTHENTICATE_FAIL_M);
+		return 	BOOT_ROM_FAIL;
+	}
+
+	boot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_SECURITY);
 
 	return BOOT_ROM_SUCCESS;
 }
@@ -275,3 +300,53 @@
 skip_bringup:
 	execute_recovery();
 }
+
+int boot_rom_verify_image(enum hl_fw_boot_ptr_imgs img_type)
+{
+	enum boot_rom_progress boot_desc_verif_base = PROGRESS_SPI_BOOT_DESC_VERIFY_START;
+	struct hl_fw_boot_ptrs boot_ptr;
+	uint32_t boot_ptr_addr = 0;
+	int ret;
+	uint32_t img_plus_key_len;
+
+	if (!security_efuse_image_validation_en())
+		return BOOT_ROM_SUCCESS;
+
+	if ((img_type == HL_FW_BOOT_PTR_SPI) && (boot_rom_global.validation_pass))
+		return BOOT_ROM_SUCCESS;
+
+	if (img_type == HL_FW_BOOT_PTR_RECOVERY)
+		boot_desc_verif_base = PROGRESS_RECOVERY_IMG_BOOT_DESC_VERIFY_START;
+
+	boot_ptr_addr = boot_desc_param_get(img_type, &boot_ptr);
+
+	if (!boot_ptr_addr)
+		return BOOT_ROM_FAIL;
+
+	boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_START);
+
+	/* TODO: once SPI DMA changes verified, check once */
+	img_plus_key_len = boot_ptr.boot_ptrs_img_sz - boot_ptr.sig_sz - CRC32_SIZE;
+
+	ret = secureboot_verify_image_signature((uint8_t *)boot_ptr_addr, img_plus_key_len,
+			(uint8_t *)(boot_ptr_addr + img_plus_key_len - boot_ptr.pub_key_sz),
+			boot_ptr.pub_key_sz, (uint8_t *)(boot_ptr_addr + img_plus_key_len), boot_ptr.sig_sz,
+			boot_ptr.sec_ver, boot_ptr.key_idx);
+
+	if (ret == SECUREBOOT_VERIFY_SUCCESS) {
+		boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_SUCCESS);
+	} else {
+		if (ret == SECUREBOOT_VERIFY_VER_FAILED)
+			boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_VERSION_FAIL);
+		else if (ret == SECUREBOOT_VERIFY_KEY_REVOKED)
+			boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_KEY_REVOKED);
+		else if (ret == SECUREBOOT_VERIFY_KEY_FAILED)
+			boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_KEY_HASH_FAIL);
+		else
+			boot_rom_status_update_bringup(boot_desc_verif_base, PROGRESS_VERIFY_AUTH_FAIL);
+
+		return BOOT_ROM_FAIL;
+	}
+
+	return BOOT_ROM_SUCCESS;
+}
diff --git a/common/secure_boot.c b/common/secure_boot.c
new file mode 100644
index 0000000..6be36ae
--- /dev/null
+++ b/common/secure_boot.c
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
+
+#include <mem_io.h>
+#include <string.h>
+#include <boot_rom.h>
+#include <dev_security.h>
+#include "secure_boot.h"
+#include <soc.h>
+#include <sha512.h>
+#include <rsa.h>
+#include <fs1_efuse.h>
+
+bool security_efuse_image_validation_en(void)
+{
+	/* check efuse for security validation enable */
+	uint32_t sig_val = boot_rom_global.sec_data.efuse_sys_cfg & 
+		EFUSE_SYS_CFG_SIGNATURE_VALIDATION_EN;
+
+	return (sig_val > 0);
+}
+
+int security_check_valid_fw_version(uint32_t fw_version)
+{
+	if ((fw_version < boot_rom_global.sec_data.sec_ver) ||
+		(fw_version > SECURITY_MAX_FW_SEC_VERSION)) {
+		return SECURITY_FW_VERSION_NOT_SECURE;
+	}
+
+	return SECURITY_FW_VERSION_SECURE;
+}
+
+int security_check_key_revoke_status(uint32_t key_index)
+{
+	uint8_t key_revoke;
+
+	if (key_index >= NUM_KEYS_IN_ROM)
+		return -1;
+
+	key_revoke = boot_rom_global.sec_data.key_revoke;
+
+	/* if the key is not revoked, return 1 */
+	if ((key_revoke & BIT(key_index)) == 0)
+		return 1;
+
+	/* no valid key found */
+	return -1;
+}
+
+int security_rom_get_key_hash_by_idx(uint32_t *hash, uint8_t key_index)
+{
+	uint32_t *ptr = (uint32_t *)rom_key_hashes[key_index];
+
+	memcpy(hash, ptr, SHA384_SUM_LEN);
+
+	return 0;
+}
+
+int secureboot_verify_key_hash(int key_index, uint8_t *key, uint32_t key_len)
+{
+	uint8_t expected_hash[SHA384_SUM_LEN];
+	uint8_t actual_hash[SHA384_SUM_LEN];
+
+	if (security_rom_get_key_hash_by_idx((uint32_t *)expected_hash, key_index))
+		return SECUREBOOT_VERIFY_KEY_FAILED;
+
+	sha384_csum(key, key_len, actual_hash, 0);
+
+	return memcmp(actual_hash, expected_hash, SHA384_SUM_LEN);
+}
+
+int secureboot_verify_image_signature(const uint8_t *image, uint32_t image_len, uint8_t *key,
+				uint32_t key_len, uint8_t *sig, uint32_t sig_len,
+				uint8_t sec_ver, uint8_t key_index)
+{
+	int ret;
+
+	if (!security_efuse_image_validation_en())
+		return SECUREBOOT_VERIFY_SUCCESS;
+
+	if ((key_len > TOTAL_PUB_KEY_PROPS_SZ) || (sig_len > SIG_SIZE))
+		return SECUREBOOT_VERIFY_FAILED;
+
+	if (security_check_valid_fw_version(sec_ver) != SECURITY_FW_VERSION_SECURE) {
+		return SECUREBOOT_VERIFY_VER_FAILED;
+	}
+
+	ret = security_check_key_revoke_status(key_index);
+	if (ret != 1)
+		return SECUREBOOT_VERIFY_KEY_REVOKED;
+
+	if (secureboot_verify_key_hash(key_index, key, key_len))
+		return SECUREBOOT_VERIFY_KEY_FAILED;
+
+	ret = rsa_verify(image, image_len, key, key_len, sig, sig_len);
+
+	/* 0 - SUCCESS, but arcboot main code, expects 1 for SUCCESS */
+	if (!ret)
+		return SECUREBOOT_VERIFY_SUCCESS;
+
+	return SECUREBOOT_VERIFY_FAILED;
+}
+
+int secureboot_verify_image_hash384(uint8_t *exp_hash, unsigned char *image, uint32_t image_len)
+{
+	uint8_t actual_hash[SHA384_SUM_LEN];
+
+	sha384_csum(image, image_len, actual_hash, 0);
+
+	return memcmp(actual_hash, exp_hash, SHA384_SUM_LEN);
+}
diff --git a/common/secure_boot.h b/common/secure_boot.h
new file mode 100644
index 0000000..7060df1
--- /dev/null
+++ b/common/secure_boot.h
@@ -0,0 +1,27 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+#ifndef _SECURE_BOOT_H_
+#define _SECURE_BOOT_H_
+
+#define SECUREBOOT_VERIFY_KEY_REVOKED	-3
+#define SECUREBOOT_VERIFY_KEY_FAILED	-2
+#define SECUREBOOT_VERIFY_VER_FAILED	-1
+#define SECUREBOOT_VERIFY_FAILED	0
+#define SECUREBOOT_VERIFY_SUCCESS	1
+
+#define SECURITY_FW_VERSION_SECURE			1
+#define SECURITY_FW_VERSION_NOT_SECURE		-1
+
+#define SECURITY_MAX_FW_SEC_VERSION			128
+
+int secureboot_verify_image_signature(const uint8_t *image, uint32_t image_len, uint8_t *key,
+				uint32_t key_len, uint8_t *sig, uint32_t sig_len,
+				uint8_t sec_ver, uint8_t key_index);
+int secureboot_verify_image_hash384(uint8_t *exp_hash, unsigned char *image, uint32_t image_len);
+
+bool security_efuse_image_validation_en(void);
+
+#endif /* _SECURE_BOOT_H_ */
diff --git a/common/sha512.c b/common/sha512.c
new file mode 100644
index 0000000..7c681cd
--- /dev/null
+++ b/common/sha512.c
@@ -0,0 +1,295 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * FIPS-180-2 compliant SHA-512 and SHA-384 implementation
+ *
+ * SHA-512 code by Jean-Luc Cooke <jlcooke@certainkey.com>
+ *
+ * Copyright (c) Jean-Luc Cooke <jlcooke@certainkey.com>
+ * Copyright (c) Andrew McDonald <andrew@mcdonald.org.uk>
+ * Copyright (c) 2003 Kyle McMartin <kyle@debian.org>
+ * Copyright (c) 2020 Reuben Dowle <reuben.dowle@4rf.com>
+ */
+
+#include <string.h>
+#include <sha512.h>
+#include <boot_rom.h>
+
+const uint8_t sha384_der_prefix[SHA384_DER_LEN] = {
+	0x30, 0x41, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86,
+	0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02, 0x05,
+	0x00, 0x04, 0x30
+};
+
+#define SHA384_H0	0xcbbb9d5dc1059ed8ULL
+#define SHA384_H1	0x629a292a367cd507ULL
+#define SHA384_H2	0x9159015a3070dd17ULL
+#define SHA384_H3	0x152fecd8f70e5939ULL
+#define SHA384_H4	0x67332667ffc00b31ULL
+#define SHA384_H5	0x8eb44a8768581511ULL
+#define SHA384_H6	0xdb0c2e0d64f98fa7ULL
+#define SHA384_H7	0x47b5481dbefa4fa4ULL
+
+static inline uint64_t Ch(uint64_t x, uint64_t y, uint64_t z)
+{
+        return z ^ (x & (y ^ z));
+}
+
+static inline uint64_t Maj(uint64_t x, uint64_t y, uint64_t z)
+{
+        return (x & y) | (z & (x | y));
+}
+
+static const uint64_t sha512_K[80] = {
+        0x428a2f98d728ae22ULL, 0x7137449123ef65cdULL, 0xb5c0fbcfec4d3b2fULL,
+        0xe9b5dba58189dbbcULL, 0x3956c25bf348b538ULL, 0x59f111f1b605d019ULL,
+        0x923f82a4af194f9bULL, 0xab1c5ed5da6d8118ULL, 0xd807aa98a3030242ULL,
+        0x12835b0145706fbeULL, 0x243185be4ee4b28cULL, 0x550c7dc3d5ffb4e2ULL,
+        0x72be5d74f27b896fULL, 0x80deb1fe3b1696b1ULL, 0x9bdc06a725c71235ULL,
+        0xc19bf174cf692694ULL, 0xe49b69c19ef14ad2ULL, 0xefbe4786384f25e3ULL,
+        0x0fc19dc68b8cd5b5ULL, 0x240ca1cc77ac9c65ULL, 0x2de92c6f592b0275ULL,
+        0x4a7484aa6ea6e483ULL, 0x5cb0a9dcbd41fbd4ULL, 0x76f988da831153b5ULL,
+        0x983e5152ee66dfabULL, 0xa831c66d2db43210ULL, 0xb00327c898fb213fULL,
+        0xbf597fc7beef0ee4ULL, 0xc6e00bf33da88fc2ULL, 0xd5a79147930aa725ULL,
+        0x06ca6351e003826fULL, 0x142929670a0e6e70ULL, 0x27b70a8546d22ffcULL,
+        0x2e1b21385c26c926ULL, 0x4d2c6dfc5ac42aedULL, 0x53380d139d95b3dfULL,
+        0x650a73548baf63deULL, 0x766a0abb3c77b2a8ULL, 0x81c2c92e47edaee6ULL,
+        0x92722c851482353bULL, 0xa2bfe8a14cf10364ULL, 0xa81a664bbc423001ULL,
+        0xc24b8b70d0f89791ULL, 0xc76c51a30654be30ULL, 0xd192e819d6ef5218ULL,
+        0xd69906245565a910ULL, 0xf40e35855771202aULL, 0x106aa07032bbd1b8ULL,
+        0x19a4c116b8d2d0c8ULL, 0x1e376c085141ab53ULL, 0x2748774cdf8eeb99ULL,
+        0x34b0bcb5e19b48a8ULL, 0x391c0cb3c5c95a63ULL, 0x4ed8aa4ae3418acbULL,
+        0x5b9cca4f7763e373ULL, 0x682e6ff3d6b2b8a3ULL, 0x748f82ee5defb2fcULL,
+        0x78a5636f43172f60ULL, 0x84c87814a1f0ab72ULL, 0x8cc702081a6439ecULL,
+        0x90befffa23631e28ULL, 0xa4506cebde82bde9ULL, 0xbef9a3f7b2c67915ULL,
+        0xc67178f2e372532bULL, 0xca273eceea26619cULL, 0xd186b8c721c0c207ULL,
+        0xeada7dd6cde0eb1eULL, 0xf57d4f7fee6ed178ULL, 0x06f067aa72176fbaULL,
+        0x0a637dc5a2c898a6ULL, 0x113f9804bef90daeULL, 0x1b710b35131c471bULL,
+        0x28db77f523047d84ULL, 0x32caab7b40c72493ULL, 0x3c9ebe0a15c9bebcULL,
+        0x431d67c49c100d4cULL, 0x4cc5d4becb3e42b6ULL, 0x597f299cfc657e2aULL,
+        0x5fcb6fab3ad6faecULL, 0x6c44198c4a475817ULL,
+};
+
+static inline uint64_t ror64(uint64_t word, unsigned int shift)
+{
+	return (word >> (shift & 63)) | (word << ((-shift) & 63));
+}
+
+#define e0(x)       (ror64(x,28) ^ ror64(x,34) ^ ror64(x,39))
+#define e1(x)       (ror64(x,14) ^ ror64(x,18) ^ ror64(x,41))
+#define s0(x)       (ror64(x, 1) ^ ror64(x, 8) ^ (x >> 7))
+#define s1(x)       (ror64(x,19) ^ ror64(x,61) ^ (x >> 6))
+
+/*
+ * 64-bit integer manipulation macros (big endian)
+ */
+#ifndef GET_UINT64_BE
+#define GET_UINT64_BE(n,b,i) {				\
+	(n) = ( (unsigned long long) (b)[(i)    ] << 56 )	\
+	    | ( (unsigned long long) (b)[(i) + 1] << 48 )	\
+	    | ( (unsigned long long) (b)[(i) + 2] << 40 )	\
+	    | ( (unsigned long long) (b)[(i) + 3] << 32 )	\
+	    | ( (unsigned long long) (b)[(i) + 4] << 24 )	\
+	    | ( (unsigned long long) (b)[(i) + 5] << 16 )	\
+	    | ( (unsigned long long) (b)[(i) + 6] <<  8 )	\
+	    | ( (unsigned long long) (b)[(i) + 7]       );	\
+}
+#endif
+#ifndef PUT_UINT64_BE
+#define PUT_UINT64_BE(n,b,i) {				\
+	(b)[(i)    ] = (unsigned char) ( (n) >> 56 );	\
+	(b)[(i) + 1] = (unsigned char) ( (n) >> 48 );	\
+	(b)[(i) + 2] = (unsigned char) ( (n) >> 40 );	\
+	(b)[(i) + 3] = (unsigned char) ( (n) >> 32 );	\
+	(b)[(i) + 4] = (unsigned char) ( (n) >> 24 );	\
+	(b)[(i) + 5] = (unsigned char) ( (n) >> 16 );	\
+	(b)[(i) + 6] = (unsigned char) ( (n) >>  8 );	\
+	(b)[(i) + 7] = (unsigned char) ( (n)       );	\
+}
+#endif
+
+static inline void LOAD_OP(int I, uint64_t *W, const uint8_t *input)
+{
+	GET_UINT64_BE(W[I], input, I*8);
+}
+
+static inline void BLEND_OP(int I, uint64_t *W)
+{
+	W[I & 15] += s1(W[(I-2) & 15]) + W[(I-7) & 15] + s0(W[(I-15) & 15]);
+}
+
+static void
+sha512_transform(uint64_t *state, const uint8_t *input)
+{
+	uint64_t a, b, c, d, e, f, g, h, t1, t2;
+
+	int i;
+	uint64_t W[16];
+
+	/* load the state into our registers */
+	a=state[0];   b=state[1];   c=state[2];   d=state[3];
+	e=state[4];   f=state[5];   g=state[6];   h=state[7];
+
+	/* now iterate */
+	for (i = 0; i < 80; i += 8) {
+		if (!(i & 8)) {
+			int j;
+
+			if (i < 16) {
+				/* load the input */
+				for (j = 0; j < 16; j++)
+					LOAD_OP(i + j, W, input);
+			} else {
+				for (j = 0; j < 16; j++) {
+					BLEND_OP(i + j, W);
+				}
+			}
+		}
+
+		t1 = h + e1(e) + Ch(e,f,g) + sha512_K[i  ] + W[(i & 15)];
+		t2 = e0(a) + Maj(a,b,c);    d+=t1;    h=t1+t2;
+		t1 = g + e1(d) + Ch(d,e,f) + sha512_K[i+1] + W[(i & 15) + 1];
+		t2 = e0(h) + Maj(h,a,b);    c+=t1;    g=t1+t2;
+		t1 = f + e1(c) + Ch(c,d,e) + sha512_K[i+2] + W[(i & 15) + 2];
+		t2 = e0(g) + Maj(g,h,a);    b+=t1;    f=t1+t2;
+		t1 = e + e1(b) + Ch(b,c,d) + sha512_K[i+3] + W[(i & 15) + 3];
+		t2 = e0(f) + Maj(f,g,h);    a+=t1;    e=t1+t2;
+		t1 = d + e1(a) + Ch(a,b,c) + sha512_K[i+4] + W[(i & 15) + 4];
+		t2 = e0(e) + Maj(e,f,g);    h+=t1;    d=t1+t2;
+		t1 = c + e1(h) + Ch(h,a,b) + sha512_K[i+5] + W[(i & 15) + 5];
+		t2 = e0(d) + Maj(d,e,f);    g+=t1;    c=t1+t2;
+		t1 = b + e1(g) + Ch(g,h,a) + sha512_K[i+6] + W[(i & 15) + 6];
+		t2 = e0(c) + Maj(c,d,e);    f+=t1;    b=t1+t2;
+		t1 = a + e1(f) + Ch(f,g,h) + sha512_K[i+7] + W[(i & 15) + 7];
+		t2 = e0(b) + Maj(b,c,d);    e+=t1;    a=t1+t2;
+	}
+
+	state[0] += a; state[1] += b; state[2] += c; state[3] += d;
+	state[4] += e; state[5] += f; state[6] += g; state[7] += h;
+
+	/* erase our data */
+	a = b = c = d = e = f = g = h = t1 = t2 = 0;
+}
+
+static void sha512_block_fn(sha512_context *sst, const uint8_t *src,
+				    int blocks)
+{
+	while (blocks--) {
+		sha512_transform(sst->state, src);
+		src += SHA512_BLOCK_SIZE;
+	}
+}
+
+static void sha512_base_do_update(sha512_context *sctx,
+					const uint8_t *data,
+					unsigned int len)
+{
+	unsigned int partial = sctx->count[0] % SHA512_BLOCK_SIZE;
+
+	sctx->count[0] += len;
+	if (sctx->count[0] < len)
+		sctx->count[1]++;
+
+	if (unlikely((partial + len) >= SHA512_BLOCK_SIZE)) {
+		int blocks;
+
+		if (partial) {
+			int p = SHA512_BLOCK_SIZE - partial;
+
+			memcpy(sctx->buf + partial, data, p);
+			data += p;
+			len -= p;
+
+			sha512_block_fn(sctx, sctx->buf, 1);
+		}
+
+		blocks = len / SHA512_BLOCK_SIZE;
+		len %= SHA512_BLOCK_SIZE;
+
+		if (blocks) {
+			sha512_block_fn(sctx, data, blocks);
+			data += blocks * SHA512_BLOCK_SIZE;
+		}
+		partial = 0;
+	}
+	if (len)
+		memcpy(sctx->buf + partial, data, len);
+}
+
+static void sha512_base_do_finalize(sha512_context *sctx)
+{
+	const int bit_offset = SHA512_BLOCK_SIZE - sizeof(uint64_t[2]);
+	uint64_t *bits = (uint64_t *)(sctx->buf + bit_offset);
+	unsigned int partial = sctx->count[0] % SHA512_BLOCK_SIZE;
+
+	sctx->buf[partial++] = 0x80;
+	if (partial > bit_offset) {
+		memset(sctx->buf + partial, 0x0, SHA512_BLOCK_SIZE - partial);
+		partial = 0;
+
+		sha512_block_fn(sctx, sctx->buf, 1);
+	}
+
+	memset(sctx->buf + partial, 0x0, bit_offset - partial);
+	/*
+	 * Original code has cpu_to_be64 instead of __builtin_bswap64.
+	 * Considering a little endian ARC system, byteswap is also a safer
+	 * alternative. This way, we need not define a new function for
+	 * cpu_to_be64.
+	 */
+
+	bits[0] = __builtin_bswap64(sctx->count[1] << 3 | sctx->count[0] >> 61);
+	bits[1] = __builtin_bswap64(sctx->count[0] << 3);
+	sha512_block_fn(sctx, sctx->buf, 1);
+}
+
+void sha384_starts(sha512_context * ctx)
+{
+	ctx->state[0] = SHA384_H0;
+	ctx->state[1] = SHA384_H1;
+	ctx->state[2] = SHA384_H2;
+	ctx->state[3] = SHA384_H3;
+	ctx->state[4] = SHA384_H4;
+	ctx->state[5] = SHA384_H5;
+	ctx->state[6] = SHA384_H6;
+	ctx->state[7] = SHA384_H7;
+	ctx->count[0] = ctx->count[1] = 0;
+}
+
+void sha384_update(sha512_context *ctx, const uint8_t *input, uint32_t length)
+{
+	sha512_base_do_update(ctx, input, length);
+}
+
+void sha384_finish(sha512_context * ctx, uint8_t digest[SHA384_SUM_LEN])
+{
+	/*
+	 * Original Code from Uboot has integer declaration for i.
+	 * This will result in a compilation warning for comparison
+	 * between signed and unsigned types. Hence replace int with
+	 * unsigned int.
+	 */
+	unsigned int i;
+
+	sha512_base_do_finalize(ctx);
+	for(i = 0; i < SHA384_SUM_LEN / sizeof(uint64_t); i++)
+		PUT_UINT64_BE(ctx->state[i], digest, i * 8);
+}
+
+/*
+ * Output = SHA-512( input buffer ).
+ *
+ * Original Code from Uboot has chunk_sz used when enabled with WD.
+ * This is intentionally removed, just like SHA256 for Preboot.
+ */
+void sha384_csum(const unsigned char *input, unsigned int ilen,
+		unsigned char *output,
+		__unused unsigned int chunk_sz)
+{
+	sha512_context ctx;
+
+	sha384_starts(&ctx);
+
+	sha384_update(&ctx, input, ilen);
+
+	sha384_finish(&ctx, output);
+}
diff --git a/include/boot_rom.h b/include/boot_rom.h
index ff9ee99..930e9fc 100644
--- a/include/boot_rom.h
+++ b/include/boot_rom.h
@@ -10,6 +10,7 @@
 #include <stdint.h>
 #include <stdbool.h>
 #include "hl_boot_desc_rom.h"
+#include <hl_sec_if.h>
 
 #define wfe()	__asm__ __volatile__("sleep 0x11\n" : : )
 
@@ -34,6 +35,7 @@
 #define SIZE_MB			(SIZE_KB * SIZE_KB)
 #define BOOT_ROM_PRIMARY				0
 #define BOOT_ROM_SECONDARY				1
+#define CRC32_SIZE		4
 
 enum boot_rom_status {
 	BOOT_ROM_SUCCESS,
@@ -96,6 +98,7 @@
 	bool fake_fuse_en;
 	uint32_t fake_fuse_spi_offset;
 	struct hl_fw_boot_ptrs spi_manifest_hdr;
+	struct hl_fw_sec_info sec_data;
 };
 
 struct boot_rom_manifest_s {
@@ -115,5 +118,6 @@
 void boot_rom_wait_for_signal(void);
 void boot_rom_check_n_enter_recovery(void);
 void boot_rom_bringup_ppboot(void);
+int boot_rom_verify_image(enum hl_fw_boot_ptr_imgs img_type);
 
 #endif /* _BOOT_ROM_H_ */
